kura/org.eclipse.kura.linux.watchdog/src/main/java/org/eclipse/kura/linux/watchdog/WatchdogServiceImpl.java

```diff
@@ -206,118 +144,149 @@ public void unregisterCriticalService(CriticalComponent criticalComponent) {
 
     @Override
     public List<CriticalComponent> getCriticalComponents() {
-        return null;
+        List<CriticalComponent> componentList = new ArrayList<CriticalComponent>();
+        for (CriticalComponentImpl cci : this.criticalServiceList) {
+            componentList.add(cci.getCriticalComponent());
+        }
+        return componentList;
     }
 
     @Override
     public void checkin(CriticalComponent criticalService) {
-        synchronized (s_criticalServiceList) {
-            for (CriticalComponentImpl csi : s_criticalServiceList) {
-                if (criticalService.getCriticalComponentName().compareTo(csi.getName()) == 0) {
-                    csi.update();
-                }
+        for (CriticalComponentImpl csi : this.criticalServiceList) {
+            if (criticalService.equals(csi.getCriticalComponent())) {
+                csi.update();
             }
         }
     }
 
-    private void doWatchdogLoop() {
-        if (this.m_enabled) {
-            if (this.m_watchdogToStop) {
-                disableWatchdog();
-                this.m_watchdogToStop = false;
-            } else {
-                boolean failure = false;
-                // Critical Services
-                synchronized (s_criticalServiceList) {
-                    if (s_criticalServiceList.size() > 0) {
-                        for (CriticalComponentImpl csi : s_criticalServiceList) {
-                            if (csi.isTimedOut()) {
-                                failure = true;
-                                s_logger.warn("Critical service {} failed -> SYSTEM REBOOT", csi.getName());
-                            }
-                        }
-                    }
-                }
-
-                if (!failure) { // refresh watchdog
-                    File f = null;
-                    FileWriter bw = null;
-                    try {
-                        f = new File(this.watchdogDevice);
-                        bw = new FileWriter(f);
-                        bw.write('w');
-                        bw.flush();
-                        s_logger.debug("watchdog refreshed");
-                    } catch (IOException e) {
-                        s_logger.error("IOException on refresh watchdog", e);
-                    } finally {
-                        try {
-                            if (bw != null) {
-                                bw.close();
-                            }
-                        } catch (IOException e) {
-                            s_logger.error("IOException on closing watchdog file", e);
-                        }
-                    }
+    private void checkCriticalComponents() {
+        // isEnabled is set if the hw watchdog device is activated
+        if (this.isEnabled) {
```

Comment:
The meaning of isEnabled and watchdogToStop are, indeed, not so clear.
isEnabled means that the hw watchdog has already been enabled and it is used to avoid that a CriticalComponent, already timed out at first run of the checkCriticalComponent, never kicks refreshWatchdog. It was already present before the refactor and I left it unchanged.
watchdogToStop, instead, is used to avoid the multiple execution of the "reboot" command.
They indeed manage corner cases and they can be replaced or removed completely.